driver-core/src/main/java/com/datastax/driver/core/Statement.java

```diff
@@ -326,4 +326,49 @@ public Statement setPagingState(PagingState pagingState) {
     ByteBuffer getPagingState() {
         return pagingState;
     }
+
+    /**
+     * Sets whether this statement is idempotent.
+     * <p>
+     * See {@link #isIdempotent()} for more explanations about this property.
+     *
+     * @param idempotent the new value.
+     * @return this {@code Statement} object.
+     */
+    public Statement setIdempotent(boolean idempotent) {
+        this.idempotent = idempotent;
+        return this;
+    }
+
+    /**
+     * Whether this statement is idempotent, i.e. whether it can be applied multiple times
+     * without changing the result beyond the initial application.
+     * <p>
+     * Idempotency plays a role in {@link com.datastax.driver.core.policies.SpeculativeExecutionPolicy speculative executions}.
+     * If a statement is <em>not idempotent</em>, the driver will not schedule speculative
+     * executions for it.
+     * <p>
+     * Note that this method can return {@code null}, in which case the driver will default to
+     * {@link QueryOptions#getDefaultIdempotency()}.
+     * <p>
+     * By default, this method returns {@code null} for all statements, except for
+     * {@link BuiltStatement}s, where the value will be inferred from the query: if it updates
+     * counters or prepends/appends to a list, the result will be {@code false}, otherwise it
+     * will be {@code true}. In all cases, calling {@link #setIdempotent(boolean)} forces a
+     * value that overrides every other mechanism.
+     *
+     * @return whether this statement is idempotent, or {@code null} to use
+     * {@link QueryOptions#getDefaultIdempotency()}.
+     */
+    public Boolean isIdempotent() {
```

Comment:
I'm slightly disturbed here by the asymmetric getter/setter pair. If idempotency is a property of statements, a statement is either idempotent or non-idempotent, so null values sound strange and conceptually would mean "I don't know if I'm idempotent" or "I may be idempotent". 
I understand that this comes from the need to introduce a "default" idempotency cluster-wide. 
But in this case the message we are sending here to the receiver slid from the initial intent: "Are you idempotent?" to: "Do you allow speculative execution?" - in which case it would be fine to return null (meaning "I don't know if I can be executed speculatively, ask the cluster").
I would suggest to make the API clearer:
* Renaming the property to `speculativeExecutionPermission` or stg in the style
* Instead of Boolean, use an enum like `SpeculativeExecutionPermission { DEFAULT, ALLOW, DENY }`